### PR TITLE
[WIP] src/tests/behave_tests: refactor behave tests

### PR DESCRIPTION
**Pull request by adk3798**
Refactoring of behave tests. Some things included:

- only create new vms on first test, re-use them for subseuent tests
- bootstrap and tear down cluster at start/end of each test
- change what options can be specified in feature description
  * can now provide bootstrap flags, ceph config options and num nodes
- check for podman and docker to use in actual kcli commands, prefer docker
- fixed basic osd test to be passing
- bootstrap no longer done using script
  * build bootstrap command using provided bootstrap args
  * execute bootstrap as generic ssh command
- added functions for running scp and local commands
  * currently scp commands used mostly for copying cephadm binary to hosts
  * local commands current only use is checking for docker/podman
- changed way ceph dev folder is handled
  * always point ceph dev folder to ceph repo where behave is being run from
  * allows any changes in python code that is mounted from ceph dev folder to
    automatically be tested in behave tests
  * should simplify using tests from developer point of view
  * before, portion of kcli template with ceph dev folder was in "raw" section
    and ceph dev folder wasn't actually mounted. Made that section not raw.
- no longer rely on kcli DNS options at all. Use explicit ips when applicable.
- avoiding ssh-copy-id because it was giving me issues, manually copying keys
- active effort to keep functions relatively small for easier maintenance

TODO:

- some way of specifying tests where it's not okay to re-use VMs. I have in
  mind tests to try ipv6 or firewall stuff where we only want that test
  worrying about it. Thinking we could tag all those tests with something
  like @special-system-settings and on those tests create fresh VMS and
  tear them down at the end.
  * maybe possible to run all these tests last so all tests that can
    re-use VMs aren't slowed down?
- Jenkins integration. Need to make sctipt that runs these tests and
  a proper Jenkins file in ceph-build repo.
- new option in steps to specifically check certain input is NOT present.
  Right now there are only positive checks
  * maybe also checks for full lines of output rather than tokenizing
    everything when it's needed (see readme, not on steps that check
    for results)
- some way to specify custom image to run tests on
- some way to specify which distro test should be run on (e.g. Fedora,
  Centos, Ubuntu)
- ~~integrate mypy checks for these files~~
- better checking for "code" value from ssh/scp commands. Right now it's
  getting code from all these commands and not checking them all

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [x] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/test/behave_tests/features/kcli_handler.py [hunk 1]
```diff
@@ -65,24 +47,45 @@ def execute_kcli_cmd(command):
...
+    """
+    try:
+        proc = subprocess.run(command.split(' '), capture_output=True, text=True, timeout=300)
+    except Exception as ex:
+        print(f"Error executing command {command}: {ex}")
+        return ('', str(ex), 1)
 
-    op = proc.stderr if proc.stderr else proc.stdout
-    return (op, proc.returncode)
+    return (proc.stdout, proc.stderr, proc.returncode)
```
Comment (sebastian-philipp) [inline src/test/behave_tests/features/kcli_handler.py]: eol

File: src/test/behave_tests/features/validation_util.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
+from typing import List
 
-def str_to_list(string):
+def str_to_list(string: str) -> List[List[str]]:
```
Comment (sebastian-philipp) [inline src/test/behave_tests/features/validation_util.py:4]: do we want to run mypy against this?

Comment (adk3798) [reply src/test/behave_tests/features/validation_util.py:4]: I think we should. I've already integrated mypy and flake8 into tox for all the files in the features directory.

**Review by sebastian-philipp — APPROVED** [review]

**Review by mgfritch — COMMENTED** [review]
can we also remove the `domain` from the kcli plan?

```
Removing old dns entry for ip 192.168.122.156
Traceback (most recent call last):
  File "/usr/bin/kcli", line 11, in <module>
    load_entry_point('kcli', 'console_scripts', 'kcli')()
  File "/root/kcli/kvirt/cli.py", line 4093, in cli
    args.func(args)
  File "/root/kcli/kvirt/cli.py", line 1641, in create_plan
    overrides=overrides, pre=pre, post=post)
  File "/root/kcli/kvirt/config.py", line 1888, in plan
    onlyassets=onlyassets)
  File "/root/kcli/kvirt/config.py", line 1007, in create_vm
    pcidevices=pcidevices, tpm=tpm, rng=rng, metadata=metadata, securitygroups=securitygroups)
  File "/root/kcli/kvirt/providers/kvm/__init__.py", line 1174, in create
    self.reserve_dns(name, nets=nets, domain=domain, alias=alias, force=True, primary=reservedns)
  File "/root/kcli/kvirt/providers/kvm/__init__.py", line 2123, in reserve_dns
    network.update(2, 10, 0, oldentry, 1)
  File "/usr/lib/python3.7/site-packages/libvirt.py", line 3311, in update
    raise libvirtError('virNetworkUpdate() failed')
libvirt.libvirtError: Operation not supported: can't update 'domain' section of network 'default'
```

**Comment by adk3798** [comment]
> can we also remove the `domain` from the kcli plan?

Done

File: src/test/behave_tests/README.md [hunk 3]
```diff
@@ -7,18 +7,16 @@ Behave framework is based on the Behaviour driven development where the test cas
...
 `features/environment.py` file is used to set up environment for testing the scenario using the kcli tool. When behave command is execute before each feature, kcli plan is generated to create the virtual machines.
 
-## Issues
-
-* We can't run the behave test cases via tox command.
-
 ## Executing the behave tests
 
-We can execute all test scenario's by executing `behave` command under `src/test/behave_test` where `features` directory is required.
+We can execute all test scenario's by executing `tox` command under `src/test/behave_test`
```
Comment (anthonyeleven) [inline src/test/behave_tests/README.md:12]: `scenarios` no \`

File: src/test/behave_tests/README.md [hunk 4]
```diff
@@ -48,3 +45,57 @@ Following implemented gherkin language steps used in `.feature` files to define
 * __I execute in {`shell`}__
 * __I wait for {`timeout`} seconds until I get__ (timeout should be defined in seconds)
 * __I get results which contain__
+
+## Note on steps that check for results in output
+
+Steps that checks for results in output do not match full strings. The output provided to check for is split up
```
Comment (anthonyeleven) [inline src/test/behave_tests/README.md:51]: s/checks/check/

File: src/test/behave_tests/README.md [hunk 5]
```diff
@@ -48,3 +45,57 @@ Following implemented gherkin language steps used in `.feature` files to define
...
+  Ex:
+      CEPH_CONFIG | mgr | mgr/cephadm/use_agent | false
+
+      would translate to
+
+      [mgr]
+      mgr/cephadm/use_agent = false
+  -------------------------------------------------------------------------
+  
+In the future, hopefully there will be an option to specify needing a fresh set of VMs or certain settings for the VMS
```
Comment (anthonyeleven) [inline src/test/behave_tests/README.md:99]: s/VMS/VMs/

You just triggered a VAX/VMS flashback : o

**Review by anthonyeleven — CHANGES_REQUESTED** [review]

**Pull request closed without merging**